warp/context.py

```diff
@@ -849,8 +849,15 @@ def __call__(self, *args, **kwargs):
 # ----------------------
 
 
-# decorator to register function, @func
-def func(f: Callable | None = None, *, name: str | None = None):
+@_overload
+def func(f: Callable, *, name: str | None = None) -> Function: ...
+@_overload
+def func(*, name: str | None = None) -> Callable[[Callable], Function]: ...
+def func(f: Callable | None = None, *, name: str | None = None) -> Any:
+    """
+    Decorator to register function, @func
+    """
+
```

Comment:
I'm afraid this is fairly confusing/overwhelming to have such stubs inline with the source code. If we want these, we would need to find a way to add them to `__init__.pyi` using the `context.export_stubs()` function.

Reply:
If there's any way we could generate these stubs automatically from the definition of these functions, then I think we'd rather take this path since it would be less error-prone and would require less maintenance in the long term. But we haven't looked into what is feasible/expected here, so I can't really comment on this.

We can always do this work later and focus this MR on the changes you've made to `def struct` and `def overload`?